Add content version management system tutorial (#18088)
cloudflare/cloudflare-docs: src/content/docs/workers/tutorials/build-a-content-version-system/index.mdx

---
updated: 2024-11-10
difficulty: Beginner
content_type: 📝Tutorial
pcx_content_type: tutorial

> title: Building a Content Version Management System with Cloudflare Workers and Durable Objects

Reviewer: Also two other questions I have:

1. Are there any restrictions on the file types this project is capable of handling for version control? For example, can it record versions for file types such as .ts, .js, .txt and so on?

2. Can it handle multiple files at a time or only one file at a time?

> title: Building a Content Version Management System with Cloudflare Workers and Durable Objects

Author: @db-cloudflare Here are my answers for your questions: 

**Question 1: Would version control system be a better term to describe the project?** => I think that it's OK, because the first thing I think about when create project is building from scratch something similar to Git using Cloudflare products. 
2. **Are there any restrictions on the file types this project is capable of handling for version control? For example, can it record versions for file types such as .ts, .js, .txt and so on?** => No, there's no limit, I just want it to have an experience similar to git diff because I like this style.
**3. Can it handle multiple files at a time or only one file at a time?** => Now it's only one file at a time now. I will continue to support diff multiple file in near future

> title: Building a Content Version Management System with Cloudflare Workers and Durable Objects

Author: @db-cloudflare  I have deployed a demo of how to utilize this guide. This is only an example, there are also other ways to use a version management: 

Management content version website: https://version-control-frontend.pages.dev/
Main site: https://content-version-system.trinhhaiyen79.workers.dev/ (Only published version are shown here) 

Hope that my example demo help you understand more about my app.

> title: Building a Content Version Management System with Cloudflare Workers and Durable Objects

Reviewer: This is amazing! Thank you very much for sharing these live examples.

The functions which we will be covering in this tutorial are as follows:
- Create a new content
- Update the content
- Get the content (specific version and list of all versions)
- Revert to a specific version
- Publish or unpublish a version
- Detailed history of modifications
- Tag a version
- Delete a version (in case you do not need it anymore)

## Prerequisites

To complete this tutorial, you only neeed:

<Render file="prereqs" product="workers" />

- Cloudflare account (Paid plan so you can use Durable Objects)
- Wrangler CLI installed on your machine
- Node.js installed on your machine
- Basic knowledge of TypeScript

If you have not already installed needed tools, you can follow the instructions in the part 0. Install tools below.

## 0. Install tools.

### Install Node.js
Node.js [official website](https://nodejs.org/en/download/package-manager)

```sh
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.0/install.sh | bash
nvm install 22
node -v # print node.js version
```

> ### Install Wrangler CLI
> 
> ```sh
> npm install -g wrangler
> ```

Reviewer: ```suggestion
```

Reviewer: We generally don't recommend installing wrangler globally to prevent incompatibilities between different projects that may require different versions of wrangler. Using C3 will install the latest version of wrangler locally to the worker project.

        "module": "ES2022",
        "moduleResolution": "node",
        "types": ["@cloudflare/workers-types"],
        "baseUrl": ".",
        "paths": {
            "@/*": ["./src/*"]
        },
        "resolveJsonModule": true,
        "allowJs": true,
        "checkJs": false,
        "noEmit": true,
        "isolatedModules": true,
        "allowSyntheticDefaultImports": true,
        "forceConsistentCasingInFileNames": true,
        "strict": true,
        "skipLibCheck": true
    },
    "include": ["src/**/*", "worker-configuration.d.ts"]
}
```

Now, we will create code files inside the 'src' directory to implement out functions. File structute will be as follows:

```sh
.src
		└───types.ts     <--- Interface definitions for data structures
		└───index.ts     <--- Entry point for request routing and presentation
		└───contentDO.ts <--- Core business logic
```
**src/types.ts: Interface definitions for data structures**
In this file, we will define data structures that will be used in the Durable Objects and the Worker. The purpose is to make sure that the data structures are consistent and more efficiently state management.

Detailed code here:

<details>

> <summary>src/types.ts</summary>

Reviewer: The `<details>` dropdown is better suited for secondary information that is not essential to complete the tutorial. For long code blocks that are required for the tutorial, I recommend taking them out of the dropdowns and try to do either of the following:

A: Use a paragraph to describe in detail the purpose for the functions within a code block

or 

B: For very long code blocks, you can break them down into sub steps where you can describe the purpose of that individual step. You can add TODO comments where a block of code requires additional functions added in a future step to help readers not get lost across the multiple steps either.

Author: @db-cloudflare Updated, please review again and let me know if there is anything that I should improve. Thank you for your help!

- `@typescript-eslint/parser` - Parser for TypeScript files in ESLint
- `eslint` - JavaScript/TypeScript linter
- `prettier` - Code formatter
- `typescript` - TypeScript compiler and language tools
- `vitest` - Unit testing framework
- `wrangler` - Cloudflare Workers CLI and development tool

Replace the `package.json` with following content:

```json
{
  "name": "content-version-system",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "deploy": "wrangler deploy",
    "dev": "wrangler dev",
    "start": "wrangler dev",
    "cf-typegen": "wrangler types",
    "lint": "npx eslint . --ext .ts --fix",
    "format": "prettier --write .",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "devDependencies": {
    "@cloudflare/workers-types": "^4.20241106.0",
    "@types/diff": "^6.0.0",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "eslint": "^8.56.0",
    "prettier": "^3.2.5",
    "typescript": "^5.3.3",
    "vitest": "^1.2.2",
    "wrangler": "^3.60.3"
  },
  "dependencies": {
    "diff": "^7.0.0",
    "itty-router": "^5.0.18"
  }
}
```
Then run `npm install` to install the dependencies.

> ### Configure tsconfig.json file

Author: When I tried to deploy in the first time, I got error and have to modify. After that, I have change codes sometimes, I will test once more with default and remove if it is not needed.

  from: string;
  to: string;
  changes: {
    additions: number;
    deletions: number;
    totalChanges: number;
    timestamp: string;
  };
  patch: string;
  hunks: Array<{
    oldStart: number;
    oldLines: number;
    newStart: number;
    newLines: number;
    lines: string[];
  }>;
}

// Publishing system - Track publishing events and history
export interface PublishRecord {
  versionId: number;
  publishedAt: string;
  publishedBy: string;
}

// API response types - Standardized response structures for the API
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

export interface DeleteVersionResponse {
  success: boolean;
  message: string;
}

export interface TagResponse {
  tagName: string;
  versionId: number;
}

export interface CurrentVersionResponse {
  version: number;
  content: string | null;
}

export type VersionListItem = Omit<Version, 'content' | 'diff'>;
```

## 3. Handle requests sent to the Durable Object

Cloudflare Workers will route incoming requests to the Durable Objects instance. The Durable Objects instance then processes the request, and provides the appropriate response to the Worker.

> To handle incoming requests, add the following code to `src/index.ts`:

Author: Replace the code with my code. I will update here to clarify

            font-family: system-ui, sans-serif;
            max-width: 800px;
            margin: 0 auto;
            padding: 20px;
        }
        .content {
            background: #f5f5f5;
            padding: 20px;
            border-radius: 8px;
            margin: 20px 0;
            white-space: pre-wrap;
        }
        .button {
            background: #0070f3;
            color: white;
            padding: 12px 24px;
            border-radius: 5px;
            text-decoration: none;
            display: inline-block;
            margin-top: 20px;
        }
        .button:hover {
            background: #0051a2;
        }
        pre {
            white-space: pre-wrap;
            word-wrap: break-word;
        }
        .meta {
            color: #666;
            font-size: 0.9em;
            margin-bottom: 10px;
        }
    </style>
</head>
<body>
    <h1>Content Version System</h1>
    <div class="content">
        <div class="meta">
            <strong>Message:</strong> ${message}<br>
            <strong>Last Updated:</strong> ${new Date(timestamp).toLocaleString()}
        </div>
        <h2>Current Content:</h2>
        <pre>${content}</pre>
    </div>

>     <a href="http://localhost:3000" class="button">Content Version Management</a>

Author: Actually because of the fact that the tutorial is quite long, I  have remove the front end code, and add the link link to front end repo in the README of backend repo: https://github.com/shinchan79/content-version-system/blob/main/README.md. Maybe I will remove this button if it makes reader confused

      fromVersion.content || '',
      toVersion.content || '',
      `Version ${fromVersion.id} (${fromVersion.message})`,
      `Version ${toVersion.id} (${toVersion.message})`
    )
  ].join('\n');

  return new Response(formattedDiff, {
    headers: {
      'Content-Type': 'text/plain',
      'Access-Control-Allow-Origin': '*',
    }
  });
}

async revertTo(versionId: number): Promise<Version> {
  const data = await this.initialize();
  const targetVersion = data.versions.find(v => v.id === versionId);
  if (!targetVersion) {
    throw new Error("Version not found");
  }

  const newVersion: Version = {
    id: this.getNextVersionId(data),
    content: targetVersion.content,
    timestamp: new Date().toISOString(),
    message: `Reverted to version ${versionId}`,
    status: targetVersion.status,
    diff: this.calculateDetailedDiff(
      data.versions[data.versions.length - 1]?.content || '',
      targetVersion.content
    )
  };

  data.versions.push(newVersion);
  await this.state.storage.put("content", data);
  return newVersion;
}
```
## 5. Testing and Deployment

You can test your project locally using Wrangler:

```sh
npx wrangler dev
```

> In your browser, navigate to `http://localhost:8787` to preview your project.

Author: navigate to `http://localhost:8787` => They might just follow the link shown in the Terminal, or copy the link and paste into browser, and make CURL is also OK, just a step to confirm that the backend code works.

To create a Worker project that is ready to use Durable Objects:

Create a Worker named `content-version-system` by running:

<PackageManagers type="create" pkg="cloudflare@latest" args={"content-version-system"} />

<Render file="c3-post-run-steps" product="workers" params={{ category: "hello-world", type: "Hello World Worker Using Durable Objects", lang: "TypeScript", }} />

Change into your new project directory:

```sh frame="none"
cd content-version-system
```

### Configure the Worker and Durable Object binding

Once `C3` has finished running, you should now have a Workers project with the name `content-version-system` that contains Cloudflare's development tool [Wrangler](/workers/wrangler/) along following files and directories:

```sh
~/content-version-system/content-version-system# ls

node_modules  package.json  package-lock.json  src  tsconfig.json  worker-configuration.d.ts  wrangler.toml
```
Update the values in your `wrangler.toml` configuration file to match the following:

```toml title="wrangler.toml"
name = "content-version-system"
main = "src/index.ts"
compatibility_date = "2024-11-06"

[observability]
enabled = true

[placement]

> mode = "smart"

Author: @harshil1712  I have read all your comment. Thanks for helping me write a better article. Please give me some days, I will try my best to resolves all the comments next week. Thank you so much!

> mode = "smart"

Author: I will also edit a little bit in the front-end repo so that the code is more understandable, and will insert the link in the tutorial for whoever need frontend part.

     - Current active version
     - List of all versions
     - Tag mappings
     - Current content
     - Publishing history

4. **Change Tracking (ContentDiff)**
   - Detailed diff information between versions
   - Tracks additions, deletions, and total changes
   - Includes patch information for showing differences

5. **Publishing System (PublishRecord)**
   - Records when and who published versions
   - Maintains an audit trail of publishing events

6. **API Responses**
   - Standardized response structures
   - Includes success/error information
   - Type-safe response data

## 3. Handle requests sent to the Durable Object

Cloudflare Workers will route incoming requests to the Durable Objects instance. The Durable Objects instance then processes the request, and provides the appropriate response to the Worker.

First, generate TypeScript types for your Worker bindings:

```sh
npm run cf-typegen
```

This command will update your `worker-configuration.d.ts` file with the correct type definitions for your Durable Object bindings.

To handle incoming requests, replace the code in `src/index.ts` with the following code:

### Initial setup and HTML template

```typescript

> import { ContentDO } from './contentDO';

Reviewer 2: We haven't defined the DO class. This will throw an error. We should move the DO code before this, or add note letting the user know that an error is expected.

- A clean HTML interface for viewing published content
- Basic styling for better readability

### CORS and helper functions

Set up CORS headers to allow cross-origin requests, then add a function to retrieve the latest published version of content from durable objects. Add the following code to `src/index.ts`:

```typescript
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, HEAD, POST, PUT, OPTIONS, DELETE',
  'Access-Control-Allow-Headers': 'Content-Type',
};

// Helper function to get latest published version from Durable Objects
async function getLatestPublishedVersion(contentDO: DurableObjectStub, origin: string): Promise<Version | null> {
  const versionsResponse = await contentDO.fetch(`${origin}/content/default/versions`);
  const versions: Version[] = await versionsResponse.json(); // Explicitly type the response

  const publishedVersions = versions.filter(v => v.status === 'published');
  if (publishedVersions.length === 0) {
    return null;
  }

  return publishedVersions.reduce((latest, current) => 
    latest.id > current.id ? latest : current
  );
}

// TODO: Request handler
```

Key features:

> - Sets up CORS headers for cross-origin requests

Reviewer 2: Why do we need CORS? Since we are not calling the worker from another URL, this shouldn't be required.

- Operations occur in the following order: read the state → process changes → write the changes.
- All operations are asynchronous because of storage communications
- There are 2 important things you have to pay attention to: First, you need to setup the CORS headers to allow cross-origin requests. Second, the class name of the Durable Object must match the name defined in the wrangler.toml file (Otherwise, error when you run `wrangler deploy` command).

### Initial setup and class definition

```typescript
import { createPatch } from 'diff';
import { 
  ContentDiff, 
  ContentState, 
  Version, 
  PublishRecord, 
  VersionStatus, 
  Tag, 
  CreateVersionRequest,
  PublishVersionRequest,
  CreateTagRequest,
  UpdateTagRequest,
  RevertVersionRequest } from './types';

// Note that the name of the class have to match the name of the Durable Object defined in the wrangler.toml file.
export class ContentDO {
  private state: DurableObjectState;
  private env: any;

  constructor(state: DurableObjectState, env: any) {
    this.state = state;
    this.env = env;
  }
  // TODO: Core Version Control Functions
}
```

This section:

> - Sets up necessary imports and CORS headers

Reviewer 2: Same as the previous comment on CORS

// Publishing system - Track publishing events and history
export interface PublishRecord {
  versionId: number;
  publishedAt: string;
  publishedBy: string;
}

// API response types - Standardized response structures for the API
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

export interface DeleteVersionResponse {
  success: boolean;
  message: string;
}

export interface TagResponse {
  tagName: string;
  versionId: number;
}

export interface CurrentVersionResponse {
  version: number;
  content: string | null;
}

export type VersionListItem = Omit<Version, 'content' | 'diff'>;
```

## 3. Handle requests sent to the Durable Object

Cloudflare Workers will route incoming requests to the Durable Objects instance. The Durable Objects instance then processes the request, and provides the appropriate response to the Worker.

To handle incoming requests, add the following code to `src/index.ts`:

### Initial Setup and HTML Template
```
import { ContentDO } from './contentDO';

interface Env {
  CONTENT: DurableObjectNamespace;
}

// HTML Template & Styling. This is used to display the published content when the user navigates to the root URL.

> // For full HTML template and styling details, see the GitHub repository.

Author: @db-cloudflare  I have added all the necessary code and verify again the correctness of the code. As long as you implement all the code in: Request handling and routing, Core version control functions, Tag management functions, Publishing and unpublishing functions, Diff and version control operations inside the ContentDO class, as the note that I added in the guide: "Please note that all code snippets in the following sections are implemented inside the ContentDO class." The application should work.

> // For full HTML template and styling details, see the GitHub repository.

Author: You can refer to the deployed version: Deployment of the 1st time (The time deployed as PoC for this guide): https://content-version-system.trinhhaiyen79.workers.dev/

Re-deploy in another workspace to verify the  correctness of this guide: https://content-version-system.sycu-lee.workers.dev/

> // For full HTML template and styling details, see the GitHub repository.

Reviewer: Thank you very much for the update. Super impressive work! Once the technical review by a second reviewer is completed we can then merge this PR.

  private state: DurableObjectState;
  private env: any;

  constructor(state: DurableObjectState, env: any) {
    this.state = state;
    this.env = env;
  }
}
```

### Request Handling and Routing

Add request handling methods:

```
// Entry point for requests to the Durable Object
async fetch(request: Request): Promise<Response> {
  if (request.method === 'OPTIONS') {
    return new Response(null, { headers: corsHeaders });
  }
  try {
    const url = new URL(request.url);
    const parts = url.pathname.split('/').filter(Boolean);
    const response = await this.handleRequest(request, parts);
    
    Object.entries(corsHeaders).forEach(([key, value]) => {
      response.headers.set(key, value);
    });

    return response;
  } catch (err) {
    const error = err as Error;
    return new Response(error.message, { status: 500, headers: corsHeaders });
  }
}

private async handleRequest(request: Request, parts: string[]): Promise<Response> {
  const path = parts.join('/');
  
  switch (`${request.method} ${path}`) {
    case 'POST content': {
      const { content, message } = await request.json();
      return Response.json(await this.createVersion(content, message));
    }

>     // Add other route handlers here (You can see the full code on GitHub repository)

Reviewer: I would add the full route handler code back here so readers can still complete a full project from reading this page alone if they want.

```sh
Ok to proceed? (y) y
In which directory do you want to create your application?
│ ./content-version-system
├ What would you like to start with?
│ Hello World example
│
├ Which template would you like to use?
│ Hello World Worker Using Durable Objects
│
├ Which language do you want to use?
│ TypeScript
├ Do you want to use git for version control?
│ yes
├ Do you want to deploy your application?
│ no
```
After done, we will have a new Workers project with the name `content-version-system` that contains following files and directories:
```sh
~/content-version-system/content-version-system# ls

node_modules  package.json  package-lock.json  src  tsconfig.json  worker-configuration.d.ts  wrangler.toml
```
We will modify the `wrangler.toml` file to specify our configuration for the Durable Objects.

```toml title="wrangler.toml"
name = "content-version-system"
main = "src/index.ts"
compatibility_date = "2024-11-06"

[observability]
enabled = true

[placement]
mode = "smart"

[[durable_objects.bindings]]
name = "CONTENT"
class_name = "ContentDO"

[[migrations]]
tag = "v1"
new_classes = ["ContentDO"]
```

> Replace the `package.json` with following content:

Reviewer: ```suggestion
### Install required packages

Replace the `package.json` with following content:
```

Reviewer: It would be good to mention which packages are being installed here.

```json
{
  "name": "content-version-system",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "deploy": "wrangler deploy",
    "dev": "wrangler dev",
    "start": "wrangler dev",
    "cf-typegen": "wrangler types",
    "lint": "npx eslint . --ext .ts --fix",
    "format": "prettier --write .",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "devDependencies": {
    "@cloudflare/workers-types": "^4.20241106.0",
    "@types/diff": "^6.0.0",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "eslint": "^8.56.0",
    "prettier": "^3.2.5",
    "typescript": "^5.3.3",
    "vitest": "^1.2.2",
    "wrangler": "^3.60.3"
  },
  "dependencies": {
    "diff": "^7.0.0",
    "itty-router": "^5.0.18"
  }
}
```
then run `npm install` to install the dependencies.

> 

Reviewer: ```suggestion

### Configure tsconfig.json file

```

> Edit the content of the `tsconfig.json` file:

Reviewer: Would be good to mention why the `tsconfig.json` file has to be edited and what part needs to be edited, even if it's just a simple package compatibility reason.

```json
{
    "compilerOptions": {
        "target": "ES2022",
        "lib": ["ES2022"],
        "jsx": "react-jsx",
        "module": "ES2022",
        "moduleResolution": "node",
        "types": ["@cloudflare/workers-types"],
        "baseUrl": ".",
        "paths": {
            "@/*": ["./src/*"]
        },
        "resolveJsonModule": true,
        "allowJs": true,
        "checkJs": false,
        "noEmit": true,
        "isolatedModules": true,
        "allowSyntheticDefaultImports": true,
        "forceConsistentCasingInFileNames": true,
        "strict": true,
        "skipLibCheck": true
    },
    "include": ["src/**/*", "worker-configuration.d.ts"]
}
```

> 

Reviewer: ```suggestion
## 2. Define the data structures for a version control system

```

Now, we will create code files inside the 'src' directory to implement out functions. File structute will be as follows:

```sh
.src
		└───types.ts     <--- Interface definitions for data structures
		└───index.ts     <--- Entry point for request routing and presentation
		└───contentDO.ts <--- Core business logic
```

You can find source code for this project on: [GitHub](https://github.com/shinchan79/content-version-system.git). Here's the explanation of the main components in this project:

> **src/types.ts: Interface definitions for data structures**

Reviewer: ```suggestion
```

}

/**
 * Step 4: Change Tracking
 * Structures for tracking differences between versions
 */
export interface ContentDiff {
  from: string;
  to: string;
  changes: {
    additions: number;
    deletions: number;
    totalChanges: number;
    timestamp: string;
  };
  patch: string;
  hunks: Array<{
    oldStart: number;
    oldLines: number;
    newStart: number;
    newLines: number;
    lines: string[];
  }>;
}

/**
 * Step 5: Publishing System
 * Track publishing events and history
 */
export interface PublishRecord {
  versionId: number;
  publishedAt: string;
  publishedBy: string;
}

/**
 * Step 6: API Response Types
 * Standardized response structures for the API
 */
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}
export interface DeleteVersionResponse {
  success: boolean;
  message: string;
}
export interface TagResponse {
  tagName: string;
  versionId: number;
}
export interface CurrentVersionResponse {
  version: number;
  content: string | null;
}
// Utility type for listing versions without full content
export type VersionListItem = Omit<Version, 'content' | 'diff'>;
```

> 

Reviewer: ```suggestion
 ## 3. Handle requests sent to the Durable Object
 
```

> **src/index.ts: Entrypoint for request routing and presentation**

Reviewer: ```suggestion
```

> In the flow of this file, Cloudflare Workers will acts as a router, route the incoming requests to the Durable Objects instance. The Durable Objects instance will process the request, and then provide the appropriate response.

Reviewer: ```suggestion
Cloudflare Workers will route incoming requests to the Durable Objects instance. The Durable Objects instance then processes the request, and provides the appropriate response to the Worker.

To handle incoming requests, add the following code to `src/index.ts`:
```

```

> /**

Reviewer: There's a lot of different functionality happening here. I would break these steps out into multiple code blocks to make it easier to digest. Each code block could be a separate tutorial step with the `###` styling for the step title (Doesn't need a number, just the title for the sub step section works). I've commented on the code that I think could be separated out into multiple code blocks and sub steps as an example.

/**
 * Step 2: HTML Template & Styling. This is used to display the published content when the user navigates to the root URL.
 * Note: For full HTML template and styling details, see the GitHub repository.
 */
const getHtmlTemplate = (content: string, message: string = '', timestamp: string = '') => `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Content Version System</title>
    <style>
        body { font-family: system-ui, sans-serif; max-width: 800px; margin: 0 auto; padding: 20px; }
        .content { background: #f5f5f5; padding: 20px; border-radius: 8px; margin: 20px 0; }
        .button { background: #0070f3; color: white; padding: 12px 24px; border-radius: 5px; }
    </style>
</head>
<body>
    <h1>Content Version System</h1>
    <div class="content">
        <div class="meta">
            <strong>Message:</strong> ${message}<br>
            <strong>Last Updated:</strong> ${new Date(timestamp).toLocaleString()}
        </div>
        <pre>${content}</pre>
    </div>
    <a href="http://localhost:3000" class="button">Content Version Management</a>
</body>
</html>
`;

> 

Reviewer: Initial setup and HTML could be step 1

/**
 * Step 3: CORS headers
 */
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, HEAD, POST, PUT, OPTIONS, DELETE',
  'Access-Control-Allow-Headers': 'Content-Type',
};

/**
 * Step 4: Version Management Helper
 * Gets the latest published version of content
 */
// Helper function to get latest published version from Durable Objects
async function getLatestPublishedVersion(contentDO: DurableObjectInstance, origin: string) {
  const versionsResponse = await contentDO.fetch(new Request(origin + '/content/default/versions'));
  const versions = await versionsResponse.json();
	 // Filter published versions and get the latest one
  const publishedVersions = versions.filter((v: any) => v.status === 'published');
  return publishedVersions.length ? 
    publishedVersions.reduce((latest: any, current: any) => latest.id > current.id ? latest : current) : 
    null;
}

> export { ContentDO };

Reviewer: CORS headers and version management helper could be step 2

        
        if (!latestPublished) {
          return new Response(JSON.stringify({ error: 'No published content' }), {
            status: 404,
            headers: { 'Content-Type': 'application/json', ...corsHeaders }
          });
        }

        const contentResponse = await contentDO.fetch(
          new Request(`${url.origin}/content/default/versions/${latestPublished.id}`)
        );
        return new Response(await contentResponse.text(), {
          headers: { 'Content-Type': 'application/json', ...corsHeaders }
        });
      }

      // Forward other requests to Durable Objects
      const response = await contentDO.fetch(request);
      const newResponse = new Response(response.body, response);
      Object.entries(corsHeaders).forEach(([key, value]) => {
        newResponse.headers.set(key, value);
      });
      return newResponse;

    } catch (error) {
      console.error('Worker error:', error);
      return new Response(JSON.stringify({
        error: 'Internal Server Error',
        message: error.message
      }), {
        status: 500,
        headers: { 'Content-Type': 'application/json', ...corsHeaders }
      });
    }
  }

> };

Reviewer: Request handling could be step 3

```

> **src/contentDO.ts: The core business logic**

Reviewer: ```suggestion
## 4. Build the core version control system logic
```

> This file allow you to perform Create - Update - Delete - Read operations  (CRUD) operations on the content versions and the version tags. Additionally, it is possible to either publish or unpublish a version, obtain detailed information on the modification history, compare versions (similar to the git `diff command`), and go back to a particular version.
> - All the states are recorded in the Durable Object storage overriding the necessity of instantiating a database instance.
> - Flows of operation: reading the state → making changes → writing the changes

Reviewer: ```suggestion
This file allow you to perform Create - Update - Delete - Read (CRUD) operations on the content versions and the version tags. Additionally, it is possible to either publish or unpublish a version, obtain detailed information on the modification history, compare versions (similar to the git `diff command`), and go back to a particular version.
- All the states are recorded in the Durable Object storage.
- Operations occur in the following order: read the state → process changes → write the changes.
```

- All operations are asynchronous because of storage communications
- There are 2 important things you have to pay attention to: First, you need to setup the CORS headers to allow cross-origin requests. Second, the class name of the Durable Object must match the name defined in the wrangler.toml file (Otherwise, error when you run `wrangler deploy` command).

> 

Reviewer: Similar to my above suggestion about separating out the `index.ts` code block, the `contentDO.ts` code block could be seperated out as well into multiple sub steps and code blocks. See my suggestions on separating it out below.

> ```

Reviewer: ```suggestion
To handle version control functions, add the following code to `src/contentDO.ts`:
```
```

/**
 * Step 1: Initial Setup & CORS Configuration
 * You will need to set up CORS headers (cross-origin requests) for your application to work properly with Durable Objects.
 */
import { createPatch } from 'diff';
import { ContentDiff, ContentState, Version, PublishRecord, VersionStatus, Tag } from './types';

const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
};

/**
 * Step 2: Durable Object Class Setup
 * Note: Note that the name of the class have to match the name of the Durable Object defined in the wrangler.toml file.
 */
export class ContentDO {
  private state: DurableObjectState;
  private env: any;
	// Durable Objects state instance and environment variables are passed to the constructor
  constructor(state: DurableObjectState, env: any) {
    this.state = state;
    this.env = env;
  }

> 

Reviewer: Initial setup, CORS configuration and Durable Object Class setup could be step 1.

        return Response.json(await this.getTags());
      }
      case `PUT content/versions/tags/${parts[3]}`: {
        const { newName } = await request.json();
        return Response.json(await this.updateTag(parts[3], newName));
      }
      // Publishing Routes
      case `POST content/default/versions/${parts[3]}/publish`: {
        const { publishedBy } = await request.json();
        return Response.json(await this.publishVersion(parseInt(parts[3]), publishedBy));
      }
      case 'GET content/default/publish-history': {
        return Response.json(await this.getPublishHistory());
      }
      // Version Control Routes 
      case `GET content/default/versions/${parts[3]}/diff`: {
        const compareToId = parseInt(new URL(request.url).searchParams.get('compare') || '0');
        return await this.getDiff(parseInt(parts[3]), compareToId || parseInt(parts[3]) - 1);
      }
      case `POST content/default/revert`: {
        const { versionId } = await request.json();
        return Response.json(await this.revertTo(versionId));
      }
      default:
        return new Response('No route matched', { status: 404 });
    }
  }

> 

Reviewer: Request handling and route management could be step 2

    };
    
    data.versions.push(newVersion);
    data.currentVersion = newVersion.id;
    data.content = content;

    await this.state.storage.put("content", data);
    return newVersion;
  }

  async getVersion(id: number): Promise<Version | null> {
    const data = await this.initialize();
    return data.versions.find(v => v.id === id) || null;
  }

  async getVersions(): Promise<Version[]> {
    const data = await this.initialize();
    return data.versions;
  }

  async deleteVersion(id: number): Promise<{ success: boolean; message: string }> {
    const data = await this.initialize();
    const versionIndex = data.versions.findIndex(v => v.id === id);
    if (versionIndex === -1) throw new Error("Version not found");

    data.versions.splice(versionIndex, 1);
    if (data.currentVersion === id) {
      const lastVersion = data.versions[data.versions.length - 1];
      data.currentVersion = lastVersion ? lastVersion.id : 0;
      data.content = lastVersion ? lastVersion.content : null;
    }

    await this.state.storage.put("content", data);
    return { success: true, message: `Version ${id} deleted successfully` };
  }

> 

Reviewer: Core version control functions could be step 3

    if (!version) throw new Error("Version not found");
    if (data.tags[name]) throw new Error("Tag exists");

    const newTag: Tag = {
      name,
      versionId,
      createdAt: new Date().toISOString()
    };
    data.tags[name] = newTag;
    await this.state.storage.put("content", data);
    return newTag;
  }

  async getTags(): Promise<Tag[]> {
    const data = await this.initialize();
    return Object.values(data.tags);
  }
  async getVersionTags(versionId: number): Promise<Tag[]> {
    const data = await this.initialize();
    return Object.values(data.tags).filter(tag => tag.versionId === versionId);
  }

  async updateTag(oldName: string, newName: string): Promise<Tag> {
    const data = await this.initialize();
    const oldTag = data.tags[oldName];
    if (!oldTag) throw new Error("Tag not found");
    if (oldName !== newName && data.tags[newName]) throw new Error("New tag exists");

    const updatedTag = {
      ...oldTag,
      name: newName,
      updatedAt: new Date().toISOString()
    };
    delete data.tags[oldName];
    data.tags[newName] = updatedTag;
    await this.state.storage.put("content", data);
    return updatedTag;
  }

> 

Reviewer: Version tag management could be step 4

  }

  async getPublishHistory(): Promise<PublishRecord[]> {
    const data = await this.initialize();
    return data.publishHistory || [];
  }
  // Unpublish a version
  async unpublishVersion(versionId: number): Promise<Version> {
    const data = await this.initialize();
    const version = data.versions.find(v => v.id === versionId);
    if (!version) {
      throw new Error("Version not found");
    }
    data.versions = data.versions.map(v => ({
      ...v,
      status: v.id === versionId ? VersionStatus.DRAFT : v.status
    }));
    if (data.publishHistory) {
      data.publishHistory = data.publishHistory.filter(
        record => record.versionId !== versionId
      );
    }
    if (data.currentVersion === versionId) {
      data.currentVersion = 0;
      data.content = null;
    }
    await this.state.storage.put("content", data);
    const updatedVersion = data.versions.find(v => v.id === versionId);
    if (!updatedVersion) {
      throw new Error("Failed to get updated version");
    }
    return updatedVersion;
  }
  async getPublishHistory(): Promise<PublishRecord[]> {
    const data = await this.initialize();
    return data.publishHistory || [];

>   }

Reviewer: Publishing system could be step 5

      },
      patch,
      hunks: []
    };
  }

  async getDiff(fromId: number, toId: number): Promise<Response> {
    const data = await this.initialize();
    const fromVersion = data.versions.find(v => v.id === fromId);
    const toVersion = data.versions.find(v => v.id === toId);
    if (!fromVersion || !toVersion) throw new Error("Version not found");

    // See GitHub repository for complete diff formatting implementation
    return new Response(createPatch('content.txt',
      fromVersion.content || '',
      toVersion.content || '',
      `Version ${fromVersion.id}`,
      `Version ${toVersion.id}`
    ), {
      headers: {
        'Content-Type': 'text/plain',
        'Access-Control-Allow-Origin': '*',
      }
    });
  }

  /**
   * Step 9: Version Control Operations
   * Additional version control features
   */
  async revertTo(versionId: number): Promise<Version> {
    const data = await this.initialize();
    const targetVersion = data.versions.find(v => v.id === versionId);
    if (!targetVersion) throw new Error("Version not found");
    return this.createVersion(targetVersion.content, `Reverted to version ${versionId}`);
  }

> }

Reviewer: Diff handling and version control operations could be step 6

  -d '{"publishedBy": "test-user"}' \
  https://<your-worker-url>/content/default/versions/{id}/publish

# View publication history
curl -X GET \
  https://<your-worker-url>/content/default/publish-history

# View diff between versions
# Option 1: E.g. you want to diff from version 2 to version 8
curl -X GET \
  https://<your-worker-url>/content/default/versions/8/diff?compare=13

# Option 2: E.g. you want to diff  8 to version 2
curl -X GET \
  https://<your-worker-url>/content/default/versions/2/diff?compare=4

#### CRUD operations for tags: ####
# 1. Create a new tag
curl -X POST "<your-worker-url>/content/versions/tags" \
  -H "Content-Type: application/json" \
  -d '{
    "versionId": 1,
    "name": "v1.0"
  }'

# 2. Get all tags
curl -X GET "https://<your-worker-url>/content/versions/tags"

# 3. Get tags for specific version
curl -X GET "https://<your-worker-url>/content/versions/1/tags"

# 4. Update tag name
curl -X PUT "https://content-version-system.trinhhaiyen79.workers.dev/content/versions/tags/v1.0" \
  -H "Content-Type: application/json" \
  -d '{
    "newName": "stable"
  }'

# 5. Delete tag
curl -X DELETE "https://<your-worker-url>/content/versions/tags/stable"
```

> ## 4. Additional tutorial tips

Reviewer: ```suggestion
## 4. Additional resources
```
